pkg/metrics/query.go

```diff
@@ -0,0 +1,188 @@
+package metrics
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	kapierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/prometheus/common/model"
+)
+
+const (
+	serviceCAinjectionDataKey             = "service-ca.crt"
+	serviceCAinjectCABundleAnnotationName = "service.beta.openshift.io/inject-cabundle"
+	metricsBaseURL                        = "https://prometheus-k8s.openshift-monitoring.svc:9091"
+)
+
+type PrometheusClient struct {
+	Client      *http.Client
+	URL         string
+	BearerToken string
+}
+
+func NewPrometheusClient(caBundle []byte, query, bearerToken string) (*PrometheusClient, error) {
+	url := fmt.Sprintf("%s/api/v1/query?%s", metricsBaseURL, (url.Values{"query": []string{query}}).Encode())
+	certPool := x509.NewCertPool()
+	certPool.AppendCertsFromPEM(caBundle)
+	tlsConfig := &tls.Config{
+		RootCAs: certPool,
+	}
+
+	tr := &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}
+	tlsClient := &http.Client{Transport: tr}
+	return &PrometheusClient{
+		Client:      tlsClient,
+		URL:         url,
+		BearerToken: bearerToken,
+	}, nil
+}
+
+func LocatePrometheus(client *kubernetes.Clientset) (string, bool) {
+	_, err := client.CoreV1().Services("openshift-monitoring").Get("prometheus-k8s", metav1.GetOptions{})
+	if kapierrs.IsNotFound(err) {
+		return "", false
+	}
+
+	var bearerToken string
+	err = wait.Poll(time.Second*1, time.Second*30, func() (bool, error) {
```

Comment:
The following comments are relevant to all uses in this PR of `wait.Poll`:

- `wait.Poll` waits for the interval before the first invocation. `wait.PollImediate` appears to be preferable in this PR so that the first invocation is executed immediately. 

- Please follow the advice @sttts provided in https://github.com/openshift/cluster-kube-controller-manager-operator/pull/316 regarding timeout.

- Suggest logging recoverable errors (i.e. via klog at warning or info level) rather than returning them to ensure that polling can continue. Only fatal errors should be returned to a `wait.Poll*` method since any non-nil error will trigger exit. 

- Note that this strategy often means that a polling method only ever reports a timeout error. Diagnosing the reason for a timeout failure then requires reviewing the logs for an indication of the recoverable error(s). In cases where more detail needs to appear directly (not necessarily in this PR), the last recoverable error can be tracked and returned instead of the timeout error.

Reply:
updated.  yup ^^ figured that all out through much trial and error